crates/miden-block-prover/src/local_block_prover.rs

```diff
@@ -0,0 +1,301 @@
+        );
+    }
+
+    Ok((nullifiers, partial_nullifier_tree.root()))
+}
+
+/// Adds the commitment of the previous block header to the chain MMR to compute the new chain root.
+fn compute_chain_root(mut chain_mmr: ChainMmr, prev_block_header: BlockHeader) -> Digest {
+    // SAFETY: This does not panic as long as the block header we're adding is the next one in the
+    // chain which is validated as part of constructing a `ProposedBlock`.
+    chain_mmr.add_block(prev_block_header, true);
+    chain_mmr.peaks().hash_peaks()
+}
+
+/// Computes the new account tree root after the given updates.
+///
+/// It uses a PartialMerkleTree for now while we use a SimpleSmt for the account tree. Once that is
+/// updated to an Smt, we can use a PartialSmt instead.
+fn compute_account_root(
+    updated_accounts: &mut Vec<(AccountId, AccountUpdateWitness)>,
+    prev_block_header: &BlockHeader,
+) -> Result<Digest, ProvenBlockError> {
+    let mut partial_account_tree = PartialMerkleTree::new();
+
+    // First reconstruct the current account tree from the provided merkle paths.
+    for (account_id, witness) in updated_accounts.iter_mut() {
+        let account_leaf_index = LeafIndex::from(*account_id);
+        // Shouldn't the value in PartialMerkleTree::add_path be a Word instead of a Digest?
+        // PartialMerkleTree::update_leaf (below) takes a Word as a value, so this seems
+        // inconsistent.
+        partial_account_tree
+            .add_path(
+                account_leaf_index.value(),
+                witness.initial_state_commitment(),
+                // We don't need the merkle path later, so we can take it out.
+                core::mem::take(witness.initial_state_proof_mut()),
+            )
+            .map_err(|source| ProvenBlockError::AccountWitnessRootMismatch {
+                account_id: *account_id,
+                source,
+            })?;
+    }
+
+    // Check the account tree root in the previous block header matches the reconstructed tree's
+    // root.
+    if prev_block_header.account_root() != partial_account_tree.root() {
+        return Err(ProvenBlockError::StaleAccountTreeRoot {
+            prev_block_account_root: prev_block_header.account_root(),
+            stale_account_root: partial_account_tree.root(),
+        });
+    }
+
+    // Second, update the account tree by inserting the new final account state commitments to
+    // compute the new root of the account tree.
+    for (account_id, witness) in updated_accounts {
+        let account_leaf_index = LeafIndex::from(*account_id);
+        partial_account_tree
+            .update_leaf(account_leaf_index.value(), Word::from(witness.final_state_commitment()))
+            .expect("every account leaf should have been inserted in the first loop");
+    }
+
+    Ok(partial_account_tree.root())
+}
+
+/// Compute the block note tree from the output note batches.
+fn compute_block_note_tree(output_note_batches: &[OutputNoteBatch]) -> BlockNoteTree {
```

Comment:
As mentioned in one of the previous comments, `output_note_batches` may need to be a different data structure (e.g., `BTreeMap` or maybe a slice with "holes") to make sure note erasure does not change the order of other notes.